RecoLocalCalo/HcalRecAlgos/interface/DoMahiAlgo.h

```diff
@@ -0,0 +1,146 @@
+#ifndef RecoLocalCalo_HcalRecAlgos_DoMahiAlgo_HH
+#define RecoLocalCalo_HcalRecAlgos_DoMahiAlgo_HH
+
+#include "DataFormats/HcalDetId/interface/HcalSubdetector.h"
+#include "DataFormats/HcalDetId/interface/HcalDetId.h"
+#include "CalibFormats/HcalObjects/interface/HcalCalibrations.h"
+#include "CalibFormats/HcalObjects/interface/HcalCoder.h"
+#include "RecoLocalCalo/HcalRecAlgos/interface/EigenMatrixTypes.h"
+#include "DataFormats/HcalRecHit/interface/HBHEChannelInfo.h"
+
+#include "CalibCalorimetry/HcalAlgos/interface/HcalPulseShapes.h"
+#include "CalibCalorimetry/HcalAlgos/interface/HcalTimeSlew.h"
+#include "RecoLocalCalo/HcalRecAlgos/interface/PulseShapeFunctor.h"
+
+#include "Math/Functor.h"
+
+class DoMahiAlgo
+{
+ public:
+  DoMahiAlgo();
+  ~DoMahiAlgo() { };
+
+  void phase1Apply(const HBHEChannelInfo& channelData, float& reconstructedEnergy, float& reconstructedTime, float& chi2);  
```

Comment:
please make this const, same as in PulseShapeFitOOTPileupCorrection and HcalDeterministicFit.
We should avoid state changes between calls.

Reply:
@slava77 This code allocates memory from the heap and then reuses it. Reallocating the memory for every channel would be too inefficient.

Would you be happy if the wokspace objects were declared mutable? Then this function could be made const.

Reply:
From my review I got an impression that multiple data members of this algorithm can be modified by multiple methods to the point that it's hard to keep track of what stays re-initialized every call to phase1Apply and what can possibly stay from the past calls an be used in the next calls.

Perhaps an "almost const" solution would be to have data members refactored, e.g. make a state struct (even on heap) with a clear reset call at the end/entry of phase1Apply.
All other methods called by phase1Apply will be const and operate on this state struct.
This way there is no unnecessary re-allocation.
At this level then the state can be made mutable and phase1Apply called const, but that then becomes somewhat optional.